zone_affinity.go

```diff
@@ -0,0 +1,169 @@
+package cluster
+
+import (
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+)
+
+const overrideEnvVar = "AVAILABILITY_ZONE"
+
+// ZoneAffinityBalancer makes a best effort to pair up consumers with partitions
+// where the leader is in the same zone.  Doing so can help from a performance
+// perspective by minimizing round trip latency, and it can also help keep costs
+// down when running in AWS by avoid cross-AZ data transfer.  Partitions are
+// spread evenly across consumers.
+//
+// The consumers communicate their zones via UserData.  If the zone is missing
+// or cannot be determined, then the data will be empty.  The zone can also be
+// set manually using the AVAILABILITY_ZONE environment variable...this is
+// useful e.g. when doing local testing.
+type ZoneAffinityBalancer struct {
+	zone []byte
+}
+
+func NewZoneAffinityBalancer() *ZoneAffinityBalancer {
```

Comment:
How about allowing the application to pass the zone as argument here? So if it needs to pick it from somewhere that's not one of the default locations it can overwrite it?

There would be a couple of options for the API:
- expose a `Zone string` property on the `ZoneAffinityBalancer` type (this is a common idiom in Go)
- treat `""` as a way to fallback to the default zone discovery mechanism, or we make a separate construct
- have two separate constructors, one receiving the zone and the other guessing it
- export the `FindZone` function

Reply:
Cool..thanks for the feedback.  I think that makes total sense.  What do you think about the functional options pattern? e.g:

```
func WithZone(zone string) func(*ZoneAffinityBalancer) {
	return func(b *ZoneAffinityBalancer) {
		b.zone = []byte(zone)
	}
}

func NewZoneAffinityBalancer(opts ...func(*ZoneAffinityBalancer)) *ZoneAffinityBalancer {
	b := &ZoneAffinityBalancer{zone: findZone()}
	for _, opt := range opts {
		opt(b)
	}
	return b
}
```

Reply:
Yep, that’s another popular one. Maybe a bit overkill for just one option... I’d personally go for the solution that needs the least amount of code.

Reply:
How do you think it reads now?  Of all the choices, I think the exported property worked best, though it meant we have to use a sentinel value (went with `unknown`) to differentiate between the zero value and inability to determine.

Reply:
Looking great 👍